### ISSUE TITLE
Train process

### ISSUE THREAD

**Issue by Rui-Zhou-2** [HopooLinZ/DAOT#3]
Hi, I am wondering if I want to retrain the model, would it be right to run pseudo_generate first? Since the outcome is mostly wrong
And what is the setting for A2B_retrain_root and A2B_fineturn_root ?

**Comment by HopooLinZ** [comment]
Yes, you need to start by generating pseudo-labels and training an initial model using the target domain images and these pseudo-labels. Although the model at this stage may have some errors, its performance will gradually improve through later fine-tuning. Even though pseudo-labels are not entirely accurate, the model can still learn some useful information.

**Comment by Rui-Zhou-2** [comment]
Hi, may I ask how many epoches you train A2B_retrain(using the target(B) domain images and pseudo-labels and A2B_fintune(using the selected) separately? I am getting pretty bad results on A2B_retrain
IMG_85.jpg Gt 5.50 Pred 469
IMG_99.jpg Gt 10.00 Pred 504
[2024-02-13 23:26:46] [32mIntermediate result: 877.1661392405064  (Index 6)[0m
 
* MAE 877.166
 * MSE 947.284

**Comment by HopooLinZ** [comment]
> Hi, may I ask how many epoches you train A2B_retrain(using the target(B) domain images and pseudo-labels and A2B_fintune(using the selected) separately? I am getting pretty bad results on A2B_retrain IMG_85.jpg Gt 5.50 Pred 469 IMG_99.jpg Gt 10.00 Pred 504 [2024-02-13 23:26:46] �[32mIntermediate result: 877.1661392405064 (Index 6)�[0m
> 
> * MAE 877.166
> * MSE 947.284

Please check your code, especially whether you are using the correct checkpoint file. There must be an issue somewhere, as we have never encountered this situation during our training process.

**Comment by kelci2017** [comment]
Hi, I’m interested in your paper and tried to run the code. Here’s a question: the modified distribution with image regions did reduced the target errors compared with using the original distribution of the source image regions, but once I calculated the original image errors (e.g. 12 regions are combined to one target image for A2B), the error was not reduced but increased. I didn’t see anywhere you calculate the errors of the original images based on the cropped regions. Do you think anything wrong with my implementation? Thanks

**Comment by HopooLinZ** [comment]
> Hi, I’m interested in your paper and tried to run the code. Here’s a question: the modified distribution with image regions did reduced the target errors compared with using the original distribution of the source image regions, but once I calculated the original image errors (e.g. 12 regions are combined to one target image for A2B), the error was not reduced but increased. I didn’t see anywhere you calculate the errors of the original images based on the cropped regions. Do you think anything wrong with my implementation? Thanks


Thank you for your interest in our work. In fact, we did not perform operations to stitch regions into complete images. During the training phase, we only used region training. Actually, random cropping operations also exist in the image preprocessing stage of the baseline, which means, even if the regions are stitched into complete images, region training is still adopted during training. Moreover, if we forcibly stitch the regions into complete images, the quality of the pictures cannot be guaranteed, which may result in some images having only heads or bodies. This could be a reason for poor performance. Our testing has always been conducted on complete images. I'm not sure if I have correctly understood your point, so please let me know if there are any issues.

**Comment by kelci2017** [comment]
Thank you for the response. I also tried testing on the original ShanghaiB train images, the test performance was not improved. Here is what I did for A2B: 1) pseudo-label the ShanghaiB train images with the given pre-trained model (model_best_57.pth), the MAE was 18.46; 2) used the pseudo-labels to re-train the pre-trained model, the MAE was improved to 15.86; 3) slide the ShanghaiA train images and gt_show and fidt h5 to get the source regions, slide the ShanghaiB train images to get the target regions; 4) feature extract the source image regions with the pre-trained model (model_best_57.pth) and feature extract the target regions with the pseudo-retrained model; 5) oot2 to select the source image regions which has the similar distribution with the target regions; 5) get the npy of these source image regions and fine-tune the pseudo retrained model; 6) pseudo label the original ShanghaiB train images with the fine-tuned model, the MAE was reduced to 19.378. In the helpers.py, I found there is no pyramids used in the slide, it just used the slide window to simply get the cropped images from the original image, not sure if that's what you want or forgot to include the pyramids. I believe your method is decent method and there may have something wrong with my implementation, but not sure where I was wrong.

**Comment by kelci2017** [comment]
I tested on the ShanghaiB test also, the performance is similar with your paper results for the pre and re-train, with the given pre-trained model, MAE=16.063, with the pseudo-label re-trained MAE=12.415. But when tested it with the SSIM and OOT, MAE=18.465. Not sure why the SSIM and OOT not working for me.

**Comment by HopooLinZ** [comment]
Indeed, if you're testing solely with SSIM and OOT and achieving an MAE of 18.465, it might be due to a lack of sufficient data when compared to the method using pseudo-labels and training on similar regions. Using SSIM and OOT alone might simulate the target domain distribution but might still be limited by the available data volume, which could prevent achieving optimal performance. However, if SSIM and OOT are applied after the pseudo-label pre-training as per our strategy and you still observed an MAE of 18.465 in the second stage, you might want to check your code or consider lowering the learning rate in the second phase. I hope this helps!

**Comment by kelci2017** [comment]
Thank you for the fast response. I just changed the data path and didn't change any of the logic you shared. Could you share the model after OOT, if possible, like the A2B after SSIM and OOT? That's fine if it's not handy. I will try to debug the code and reproduce it.

**Comment by HopooLinZ** [comment]
This is the Baidu Cloud link for the A2B weights: https://pan.baidu.com/s/1zF8Zs1kmU14u7FUyaDE06w?pwd=49cy Access code: 49cy. Copy this section to Baidu Cloud. It is the result of a laboratory reproduction, and the best weight file was lost, with an error of about 0.5.

**Comment by kelci2017** [comment]
Thanks for the response. I'm wondering if you plan to use the pyramids in the image regions. Seems like you just cropped the original images with regions size 256 without scale. I tried the SSIM only after retraining, it selected 410 image regions from the 4800 regions, the same selections with OOT. I didn't see any performance difference as you did in the paper. I mean for the A2B.

**Comment by HopooLinZ** [comment]
In our experiments, there is a difference between SSIM and SSIM+OT because the SSIM underwent a Sinkhorn algorithm. In fact, our ablation study and Figure 8 also demonstrate the slight differences between OT and using SSIM alone. Please check if there is an issue with your code, or re-examine whether the selected areas are different. The region size we used is 256*256.

**Comment by Rui-Zhou-2** [comment]
Hi, I am also wondering if the slide and oot are conducted on gt_show?  Isn't the ground truth for the target image not accessible in the Da setting?
https://github.com/HopooLinZ/DAOT/blob/e6b8c033fff395b8588240a47251674a99e565d8/slide.py#L17

**Comment by HopooLinZ** [comment]
In the A2B setup, the ground truth of A is known.
